### PR TITLE
EL 4.0.0 part 1

### PR DESCRIPTION
**Pull request by markt-asf**
Signed-off-by: Mark Thomas <markt@apache.org>

Include the following in the PR:
- [x] A directory in the form wombat/x.y where x.y is the release major.minor version, and the directory contains the following.
- [x] Specification PDF in the form of wombat-spec-x.y.pdf
- [x] Specification HTML in the form of wombat-spec-x.y.html
- [x] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [x] For a Progress Review, that sufficient progress has been made on a Compatible Implementation and TCK, to ensure that the spec is implementable and testable. 
- [x] For a Release Review, a summary that a Compatible Implementation is complete, passes the TCK, and that the TCK includes sufficient coverage of the specification.
The TCK users guide MUST include the instructions to run the compatible implementations used to validate the release.
Instructions MAY be by reference.
- [x] The URL of the OSSRH staging repository for the api, javadoc:
      https://jakarta.oss.sonatype.org/content/groups/staging/jakarta/el/jakarta.el-api/4.0.0/
- [x] The URL of the staging directory on download.eclipse.org for the proposed EFTL TCK binary:
      https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/staged-900/jakarta-expression-language-tck-4.0.0.zip
- [x] The URL of the compatibility certification request issue:
      https://github.com/eclipse-ee4j/el-ri/issues/134

### CODE REVIEW

**Comment by markt-asf** [comment]
This is draft pending the availability of the proposed EL 4.0.0 TCK.
I am not aware of any remaining issues blocking the release of the EL 4.0.0 TCK. At the time of writing this comment, Tomcat 10.0.0-M7 passes the EL 4.0.0 TCK built from the repo with both Java 8 and Java 11.

**Comment by kwsutter** [comment]
Thanks, @markt-asf!  FYI, we're supposed to be assigning Spec Committee mentors to each of these PRs as an aid to get through the process.  Unfortunately, we've had a little blip with our own process and the mentors for EL are not official Committers yet for this repo... ;-)  So, I will take the initial look and provide comments.

**Comment by kwsutter** [comment]
Here's the overall Checklist that we will use to help ensure that EL is ready for ballot...  The mentors should be ensuring that all of these items are in place.  But, the EL team is welcome to review this checklist as well to ensure that everything is in order.  Thanks.

# Spec Review Checklist

1. Spec PR
  - [ ] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [ ] Directory of form {spec}/x.y
  - [ ] PDF of form {spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [ ] HTML of form {spec}-spec-x.y.html ("-spec" preferred but not required)
  - [ ] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [ ] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [ ] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [ ] Staging repository link of the form https://oss.sonatype.org/conte/repositories/staging/jakarta/{spec}/jakarta-{spec}-api/x.y.z/
  - [ ] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [ ] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [ ] (Optional) Second PR for just apidocs

2. _index.md
  - [ ] Link to spec pdf
  - [ ] Link to spec html
  - [ ] Link to apidocs
  - [ ] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip
  - [ ] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [ ] Name of and link to at least one Compatible Implementation

3. javadocs
  - [ ] Footer contains Eclipse copyright and link to license
  - [ ] ESFL license is included, usually as doc-files/speclicense.html
  - [ ] no META-INF directory in PR
  - [ ] javadocs-jar artifact matches apidocs (optional for this release)

4. Spec PDF
  - [ ] Correct spec title
  - [ ] Version number of the form x.y, not x.y.z
  - [ ] Correct Eclipse copyright line
  - [ ] No DRAFT or SNAPSHOT
  - [ ] Correct Logo

5. Spec HTML
  - [ ] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [ ] EFTL license file, preferably named LICENSE.md
  - [ ] User's Guide (or equivalent documentation)
  - [ ] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [ ] Software requirements listed
  - [ ] Installation and configuration described
  - [ ] How to run tests
  - [ ] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors
  - [ ] SHA-256 fingerprint matches staged TCK zip file on cert request
 
10. Update Jakarta EE API jar
  - [ ] Update the Jakarta EE API jar by submitting a PR to the jakartaee-api project that updates the version number of your API jar file.

File: expression-language/4.0/_index.md
Comment (markt-asf) [reply expression-language/4.0/_index.md]: One of my common typos. Sorry about that. Fixed.

File: expression-language/4.0/_index.md
Comment (kwsutter) [reply expression-language/4.0/_index.md:13]: FYI, here's the Issue for getting this TCK staging area setup and processed...
https://github.com/eclipse-ee4j/jakartaee-tck/issues/360

File: expression-language/4.0/_index.md
Comment (markt-asf) [reply expression-language/4.0/_index.md:20]: No strong preference. My intention was to aim to have both with the option to use just one if we needed to move forward before the other was ready for some reason.

File: expression-language/4.0/_index.md
Comment (scottmarlow) [reply expression-language/4.0/_index.md:13]: https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/staged-900/eclipse-expression-language-tck-4.0.0.zip is now available.

Comment (markt-asf) [reply expression-language/4.0/_index.md:13]: Just to confirm, that download directory contains the proposed final TCKs and (assuming no bugs) they won't change?

Comment (scottmarlow) [reply expression-language/4.0/_index.md:13]: Hmm, I will look into the "eclipse-" prefix in the file name.

Comment (scottmarlow) [reply expression-language/4.0/_index.md:13]: @markt-asf Do you expect that only final (passing 100%) TCKs would be in the staged folder?  I expect that we will keep making changes (e.g bug fixes + doc updates).  

When we have TCKs that are expected to be final, we will promote them to https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/promoted folder.

Also the file names look wrong with the "eclipse-" prefix, so I expect we will generate a "expression-language-tck-4.0.0.zip" instead (https://github.com/eclipse-ee4j/jakartaee-tck/issues/363 is for addressing the unexpected "eclipse-" prefix).

Comment (scottmarlow) [reply expression-language/4.0/_index.md:13]: I guess the "eclipse-" prefix is expected in [build_standalone-tcks.sh#L244](https://github.com/eclipse-ee4j/jakartaee-tck/blob/master/docker/build_standalone-tcks.sh#L244)

Comment (alwin-joseph) [reply expression-language/4.0/_index.md:13]: Based on the discussion in https://github.com/eclipse-ee4j/jakartaee-tck/issues/363 the eclipse-* bundles will be renamed to jakarta-*.

Also we require to update the userguides (especially the versions) for all the TCKs along with other possible changes. 

As Scott mentioned the https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/staged-900 will contain the nightly EFTL bundles. We will move them to https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/promoted which can be used to run the certification tests.

Comment (kwsutter) [reply expression-language/4.0/_index.md:13]: > Based on the discussion in [eclipse-ee4j/jakartaee-tck#363](https://github.com/eclipse-ee4j/jakartaee-tck/issues/363) the eclipse-* bundles will be renamed to jakarta-*.

Correct.  The final name should begin with `jakarta-`, but early versions in Jakarta EE 8 accidentally had the `eclipse-` prefix. Instead of forcing a change in all of the PRs, we decided to go ahead and change the name when we run the promotion script.  Going forward, we should try to use the `jakarta-` prefix.

> 
> Also we require to update the userguides (especially the versions) for all the TCKs along with other possible changes.

Correct.

Comment (alwin-joseph) [reply expression-language/4.0/_index.md:13]: 1. https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/staged-900/ now has the bundles with "jakarta-" prefix. Once the TCKs are final they will be moved to https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/promoted.

2. I raised https://github.com/eclipse-ee4j/jakartaee-tck/pull/367  for updating the EL TCK userguide. I hope each API teams can pitch in to update the respective TCK userguides. 
Ref : https://github.com/jakartaee/specifications/pull/229#issuecomment-660733738

Comment (kwsutter) [reply expression-language/4.0/_index.md:13]: This conversation looks to be resolved.

**Comment by kwsutter** [comment]
@MattGill98, can you help out with mentoring this PR across the finish line?  The TCK and CI are the main items left to verify.  Thanks!

**Comment by MattGill98** [comment]
This still needs actions as per my review, can somebody make those changes?

**Comment by alwin-joseph** [comment]
> This still needs actions as per my review, can somebody make those changes?



> This still needs actions as per my review, can somebody make those changes?

There was Userguide and doc changes made for EL tck, please suggest if it is ok to refresh the tck bundle at https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/staged-900/jakarta-expression-language-tck-4.0.0.zip so it can be reviewed.

**Comment by MattGill98** [comment]
Hi @alwin-joseph, yes please feel free to update that bundle

**Comment by MattGill98** [comment]
@alwin-joseph it looks as if you've updated the bundle - thank you for that! I'll just re-review it now...

**Comment by arjantijms** [comment]
Thanks @MattGill98! Just send me an IM if you have any quick questions.

I'm now going to update the API with 1 change, which is the automatic module that was asked for. I'll of course also need to re-run the TCK with the updated bundle and update the SHAs.

**Comment by MattGill98** [comment]
Perfect thank you @arjantijms.

It looks like the docs/el_tck_properties.txt file still has 3.0 as the short version, which looks to be the only lingering old version.

My suggested URL change still needs making I believe

**Comment by arjantijms** [comment]
>  docs/el_tck_properties.txt

I think that file might not be used anymore. There's more wrongness in it.

@alwin-joseph can we delete the `docs/el_tck_properties.txt` file?

**Comment by arjantijms** [comment]
> My suggested URL change still needs making I believe

Yes, I think only @markt-asf can make that change as he did the PR here.

**Comment by alwin-joseph** [comment]
> > docs/el_tck_properties.txt
> 
> I think that file might not be used anymore. There's more wrongness in it.
> 
> @alwin-joseph can we delete the `docs/el_tck_properties.txt` file?

Sure. https://github.com/eclipse-ee4j/jakartaee-tck/pull/460 . Please suggest if any other changes required for TCK.

**Comment by alwin-joseph** [comment]
> 
> It looks like the docs/el_tck_properties.txt file still has 3.0 as the short version, which looks to be the only lingering old version.
> 
The docs/el_tck_properties.txt file has been removed and the TCK is refreshed now.

**Comment by arjantijms** [comment]
Thank you @alwin-joseph, I'll run the TCK again.

**Comment by arjantijms** [comment]
New run (with updated TCK):

```
[javatest.batch] PASSED........com/sun/ts/tests/signaturetest/el/ELSigTest.java#signatureTest
[javatest.batch] 
[javatest.batch] Aug 20, 2020 6:15:05 PM Finished executing all tests, wait for cleanup...
[javatest.batch] Aug 20, 2020 6:15:05 PM Harness done with cleanup from test run.
[javatest.batch] Total time = 478s
[javatest.batch] Setup time = 0s
[javatest.batch] Cleanup time = 0s
[javatest.batch] Test results: passed: 336
[javatest.batch] Results written to /home/jenkins/agent/workspace/2_expression-language-run-tck-against-staged-build/work.
[javatest.batch] Report written to /home/jenkins/agent/workspace/2_expression-language-run-tck-against-staged-build/report

BUILD SUCCESSFUL
Total time: 8 minutes 0 seconds
SHA256_GF=16ffa2afdbfb880229d1b4701d0470d7617664fec4e2c898be2dc04adae9509f
SHA256_API=bdac40bf13816ce511c83fe00be4ab3447c5cd0e2dfc0c7c03d9ce8fa4b3e9e7
SHA256_TCK=fa359403e898edbcf81e01c03979e8a052c6e21fc06fc1b215078cc4a2d57d2c
Product_download=https://ci.eclipse.org/glassfish/job/glassfish_build-and-test-using-jenkinsfile/job/master/238/artifact/bundles/glassfish.zip
API_download=https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/el/jakarta.el-api/4.0.0/jakarta.el-api-4.0.0.jar
TCK_download=https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/staged-900/jakarta-expression-language-tck-4.0.0.zip
OS4=3.12.0 NAME="Alpine Linux" ID=alpine VERSION_ID=3.12.0 PRETTY_NAME="Alpine Linux v3.12" HOME_URL="https://alpinelinux.org/" BUG_REPORT_URL="https://bugs.alpinelinux.org/"
JDK_VERSION=java version "1.8.0_202" Java(TM) SE Runtime Environment (build 1.8.0_202-b08) Java HotSpot(TM) 64-Bit Server VM (build 25.202-b08, mixed mode)
```

File: expression-language/4.0/_index.md
Comment (markt-asf) [reply expression-language/4.0/_index.md]: PR updated

**Comment by MattGill98** [comment]
Thank you @markt-asf and @arjantijms for your continued work. Looks like just a certification request is needed now to mark this as complete

**Comment by alwin-joseph** [comment]
I assume that there are no more changes needed to EL TCK (https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/staged-900/jakarta-expression-language-tck-4.0.0.zip). It will be promoted to https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/promoted/jakarta-expression-language-tck-4.0.0.zip soon as described in https://github.com/eclipse-ee4j/jakartaee-tck/issues/387.

**Comment by arjantijms** [comment]
@alwin-joseph There are indeed no outstanding changes to the EL TCK that I know of.

**Comment by arjantijms** [comment]
@MattGill98 Thanks for the mentoring. The CR issue is here: 

https://github.com/eclipse-ee4j/el-ri/issues/134

**Comment by arjantijms** [comment]
The TCK results PR for GF is here: https://github.com/eclipse-ee4j/glassfish/pull/23195

**Comment by edbratt** [comment]
Please add a comment indicating you have approved the EL TCK for 
promotion so it won't be rebuilt

https://github.com/eclipse-ee4j/jakartaee-tck/issues/387

Until it is promoted out of the "Stage" directory folder, it's still 
possible something will cause it to be rebuilt and we don't want that at 
this point.

**Comment by MattGill98** [comment]
Thanks @arjantijms! One final thing I can see is needed before ballot is an API PR of the form: https://github.com/eclipse-ee4j/jakartaee-api/pull/###

**Comment by arjantijms** [comment]
@MattGill98 I almost forgot that API one, but here it is: https://github.com/eclipse-ee4j/jakartaee-api/pull/81

File: expression-language/4.0/_index.md
Comment (jeanouii) [inline expression-language/4.0/_index.md:10]: The files need the jakarta- prefix

See PR https://github.com/eclipse-ee4j/el-ri/pull/135

**Review by jeanouii — CHANGES_REQUESTED** [review]

**Comment by MattGill98** [comment]
@arjantijms or @markt-asf if you wouldn't mind addressing the requested changes, I can put this forward to ballot

**Comment by arjantijms** [comment]
@MattGill98 I can't make the changes directly as this is the PR of @markt-asf 

I can do a PR to his branch though.

File: expression-language/4.0/_index.md
Comment (markt-asf) [reply expression-language/4.0/_index.md]: Renamed as requested.

**Comment by markt-asf** [comment]
I think everything is as required now. If not, I should be able to find time to tweak things further.

**Comment by arjantijms** [comment]
@jeanouii @MattGill98 @kwsutter can you check if everything has been addressed now?

**Comment by Pandrex247** [comment]
Matt is on holiday for a couple of days so I'm just having a quick scan - I believe it's good to go though as last time we spoke those now resolved issues were the only things holding it back.

**Review by Pandrex247 — APPROVED** [review]

File: expression-language/4.0/_index.md
Comment (kwsutter) [inline expression-language/4.0/_index.md:13]: The [sig], [sha], and [pub] urls need to be filled in.  Ensure these are corrected before merging, but I wouldn't hold up the vote for these corrections.

**Review by kwsutter — COMMENTED** [review]
The url links need to be filled in for the rest of the TCK artifacts.

**Comment by hussainnm** [comment]
JSP acronym is used in a few places. As per the [Acronym Guidelines](https://jakarta.ee/legal/acronym_guidelines/), it would be better to have a line in the beginning of the document like _"References in this document to JSP refer to the Jakarta Server Pages unless otherwise noted"_

**Comment by arjantijms** [comment]
Release bug: https://bugs.eclipse.org/bugs/show_bug.cgi?id=567011
IP log bug: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22589
Ballot: https://eclipse.org/lists/jakarta.ee-spec/msg00895.html

**Comment by MattGill98** [comment]
I'm just about to conclude the ballot, @markt-asf could you add the checksums as indicated by Kevin please?

**Comment by arjantijms** [comment]
@MattGill98 Since you have to edit `_index.md` anyway to enter the votes, can you add this:

```
* [Jakarta Expression Language 4.0 TCK](https://download.eclipse.org/jakartaee/expression-language/4.0/jakarta-expression-language-tck-4.0.0.zip)([sig](https://download.eclipse.org/jakartaee/expression-language/4.0/jakarta-expression-language-tck-4.0.0.zip.sig),[sha](https://download.eclipse.org/jakartaee/expression-language/4.0/jakarta-expression-language-tck-4.0.0.zip.sha256),[pub](https://raw.githubusercontent.com/jakartaee/specification-committee/master/jakartaee-spec-committee.pub))
```

It comes in place of:

```
* [Jakarta Expression Language 4.0 TCK](https://download.eclipse.org/jakartaee/expression-language/4.0/jakarta-expression-language-tck-4.0.0.zip)([sig](),[sha](),[pub]())
```

The release review table btw should become:

```
| Representative                                      | Representative for: | Vote |
|-----------------------------------------------------|---------------------|------|
| Kenji Kazumura                                      | Fujitsu             |      |
| Dan Bandera, Kevin Sutter                           | IBM                 |      |
| Ed Bratt, Dmitry Kornilov                           | Oracle              |      |
| Andrew Pielage, Matt Gill                           | Payara              |      |
| Scott Stark, Mark Little                            | Red Hat             |      |
| David Blevins, Jean-Louis Monteiro, Cesar Hernandez | Tomitribe           |      |
| Ivar Grimstad                                       | EE4J PMC            |      |
| Marcelo Ancelmo, Martijn Verburg                    | Participant Members |      |
| Werner Keil                                         | Committer Members   |      |
| Scott (Congquan) Wang                               | Enterprise Members  |      |
|                                                     | Total               |      |
```

With the right votes filled-in of course ;)

**Comment by MattGill98** [comment]
On ballot completion, the specification committee mentor:
 - [ ] adds this final checklist to the main PR.
 - [ ] adds the `approved` label to the PRs, and sends out the Ballot Summary per this [template](#ballot-summary-email-template) to the [public Jakarta EE Specification Committee email list](jakarta.ee-spec@eclipse.org)
 - [ ] calculates the staged EFTL TCK signature and promotes it to the committee download area
  using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job. Manually editing the jenkins Build Information will help identify the build (ie. Mail 2.0 or CDI 3.0).
 - [ ] merges the specification (and apidocs) PRs, ensuring the "date:" field in the _index.md file has an appropriate value to allow publishing.
 - [ ] updates the specification page with the ballot results. This is normally done via a separate PR that should be reviewed, approved, and merged.
 - [ ] notifies the EMO of the ballot results by email to [emo@eclipse-foundation.org](emo@eclipse-foundation.org). Just forward the ballot summary note sent earlier to the public Spec Committee email list.
 - [ ] creates an issue in the specification project that includes the following checklist for the specification project team:
 \<include url of Issue here>


 ----
 - [ ] promotes api staging release promotes the specification api jars to maven central. An example release job script can be found here https://wiki.eclipse.org/MavenReleaseScript.
 - [ ] go through the merged jakarta.ee specification website page to verify all the links are valid.
 - [ ] approve the compatibility request.
 - [ ] The compatible implementation project/vendor MUST send an email to tck@eclipse.org for approval of the compatible implementation for trademark usage.
 - [ ] merge any final release branch as appropriate for the branch management for the project.

File: expression-language/4.0/_index.md
Comment (MattGill98) [reply expression-language/4.0/_index.md]: I've made the change myself, @kwsutter does that look better?

**Comment by MattGill98** [comment]
@arjantijms I've made the necessary change to this PR to prevent merging incorrect lines, and pushed the ballot results to this PR as it was one go. Do you think it would be better if I split the ballot results into a new PR?

**Comment by markt-asf** [comment]
Looks like @MattGill98 beat me to it. I'm assuming I don't need to do anything else. Ping me if that is not the case.

**Comment by arjantijms** [comment]
@MattGill98 I think all is fine now, and you can merge? You do have to resolve your own requested changes ;)

**Review by MattGill98 — APPROVED** [review]

**Comment by MattGill98** [comment]
Seemed to diff weirdly, I've merged master to fix it. Looks good to merge now!

**Pull request merged**
